lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -0,0 +1,770 @@
+//===- InferWidths.cpp - Infer width of types -------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the InferWidths pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/Hashing.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "infer-widths"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Constraint Expressions
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct Expr;
+} // namespace
+
+/// Allow rvalue refs to `Expr` and subclasses to be printed to streams.
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const T &e) {
+  e.print(os);
+  return os;
+}
+
+/// Allow pointers to `Expr` and subclasses to be printed to streams.
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const T *e) {
+  return os << *e;
+}
```

Comment:
Please pick one or the other (I'd recommend `const T&`) to eliminate the enable_if's and simplify the code.